crates/storage/src/blueprint/merklelized.rs

```diff
@@ -0,0 +1,490 @@
+    >,
+    Self: Blueprint<M, S>,
+    for<'a> StructuredStorage<&'a S>: StorageInspect<Metadata, Error = StorageError>,
+{
+    fn root(storage: &S, key: &M::Key) -> StorageResult<MerkleRoot> {
+        use crate::StorageAsRef;
+        let storage = StructuredStorage::new(storage);
+        let key = key.to_owned().into();
+        let metadata = storage
+            .storage_as_ref::<Metadata>()
+            .get(&DenseMetadataKey::Primary(key))?
+            .ok_or(not_found!(Metadata))?;
+        Ok(*metadata.root())
+    }
+}
+
+impl<M, S, KeyCodec, ValueCodec, Metadata, Nodes, Encoder> SupportsBatching<M, S>
+    for Merklelized<KeyCodec, ValueCodec, Metadata, Nodes, Encoder>
+where
+    M: Mappable,
+    S: BatchOperations,
+    KeyCodec: Encode<M::Key> + Decode<M::OwnedKey>,
+    ValueCodec: Encode<M::Value> + Decode<M::OwnedValue>,
+    Encoder: Encode<M::Value>,
+    Metadata: Mappable<
+        Key = DenseMetadataKey<M::OwnedKey>,
+        OwnedKey = DenseMetadataKey<M::OwnedKey>,
+        Value = DenseMerkleMetadata,
+        OwnedValue = DenseMerkleMetadata,
+    >,
+    Nodes: Mappable<Key = u64, Value = Primitive, OwnedValue = Primitive>,
+    for<'a> StructuredStorage<&'a mut S>: StorageMutate<Metadata, Error = StorageError>
+        + StorageMutate<Nodes, Error = StorageError>,
+{
+    fn init<'a, Iter>(storage: &mut S, column: S::Column, set: Iter) -> StorageResult<()>
+    where
+        Iter: 'a + Iterator<Item = (&'a M::Key, &'a M::Value)>,
+        M::Key: 'a,
+        M::Value: 'a,
+    {
+        <Self as SupportsBatching<M, S>>::insert(storage, column, set)
+    }
+
+    fn insert<'a, Iter>(
+        storage: &mut S,
+        column: S::Column,
+        set: Iter,
+    ) -> StorageResult<()>
+    where
+        Iter: 'a + Iterator<Item = (&'a M::Key, &'a M::Value)>,
+        M::Key: 'a,
+        M::Value: 'a,
+    {
+        for (key, value) in set {
+            <Self as Blueprint<M, S>>::replace(storage, key, column, value)?;
+        }
+
+        Ok(())
+    }
+
+    fn remove<'a, Iter>(
+        storage: &mut S,
+        column: S::Column,
+        set: Iter,
+    ) -> StorageResult<()>
+    where
+        Iter: 'a + Iterator<Item = &'a M::Key>,
+        M::Key: 'a,
+    {
+        for item in set {
+            let key_encoder = KeyCodec::encode(item);
+            let key_bytes = key_encoder.as_bytes();
+            Self::remove(storage, key_bytes.as_ref(), column)?;
+        }
+        Ok(())
+    }
+}
+
+/// The macro that generates basic storage tests for the table with the merklelized structure.
+/// It uses the [`InMemoryStorage`](crate::structured_storage::test::InMemoryStorage).
+#[macro_export]
+macro_rules! basic_merklelized_storage_tests {
+    ($table:ident, $key:expr, $value_insert:expr, $value_return:expr, $random_key:expr) => {
+        $crate::paste::item! {
+        #[cfg(test)]
+        #[allow(unused_imports)]
+        mod [< $table:snake _basic_tests >] {
```

Comment:
These look identical to the basic storage tests, should we just reuse those here?

Also, perhaps we should have some merkle tree related tests as part of this macro?

Reply:
The tests are not the same since the removal of elements is not allowed=)

I moved old tests for the `FuelBlocks` table in the `blocks.rs` instead of adding new tests here. I did it for the simplicity of the PR, but if you think it is better to make it part of this macro now, then later, I will do it=)